syntax = "proto3";

import "scanner/metadata.proto";
package scanner.proto;

service Master {
  /// Database metadata methods

  // List table info
  rpc ListTables (Empty) returns (ListTablesResult) {}
  // Get table info
  rpc GetTables (GetTablesParams) returns (GetTablesResult) {}
  // Delete table in the database
  rpc DeleteTables (DeleteTablesParams) returns (Empty) {}
  rpc NewTable (NewTableParams) returns (Empty) {}
  rpc GetVideoMetadata (GetVideoMetadataParams) returns (GetVideoMetadataResult) {}

  /// Worker registration methods
  // Called after a new worker spawns to register with the master
  rpc RegisterWorker (WorkerParams) returns (Registration) {}
  // Called when a worker is removed
  rpc UnregisterWorker (UnregisterWorkerRequest) returns (Empty) {}
  rpc ActiveWorkers (Empty) returns (RegisteredWorkers) {}
  // Ingest videos into the system
  rpc IngestVideos (IngestParameters) returns (IngestResult) {}

  // Register new methods
  rpc LoadOp (OpPath) returns (Result) {}
  rpc RegisterOp (OpRegistration) returns (Result) {}
  rpc RegisterPythonKernel (PythonKernelRegistration) returns (Result) {}

  rpc GetOpInfo (OpInfoArgs) returns (OpInfo) {}
  rpc GetSourceInfo (SourceInfoArgs) returns (SourceInfo) {}
  rpc GetSinkInfo (SinkInfoArgs) returns (SinkInfo) {}
  rpc GetEnumeratorInfo (EnumeratorInfoArgs) returns (EnumeratorInfo) {}

  rpc ListLoadedOps (Empty) returns (ListLoadedOpsReply) {}
  rpc ListRegisteredOps (Empty) returns (ListRegisteredOpsReply) {}
  rpc ListRegisteredPythonKernels (Empty) returns (ListRegisteredPythonKernelsReply) {}

  rpc Ping (Empty) returns (Empty) {}
  rpc Shutdown (Empty) returns (Result) {}
  rpc PokeWatchdog (Empty) returns (Empty) {}

  // Internal
  rpc NextWork (NextWorkRequest) returns (NextWorkReply) {}
  rpc FinishedWork (FinishedWorkRequest) returns (Empty) {}
  rpc FinishedJob (FinishedJobRequest) returns (Empty) {}
  rpc NewJob (BulkJobParameters) returns (NewJobReply) {}

  rpc GetJobs (GetJobsRequest) returns (GetJobsReply) {}
  rpc GetJobStatus (GetJobStatusRequest) returns (GetJobStatusReply) {}

}

service Worker {
  rpc NewJob (BulkJobParameters) returns (Result) {}
  rpc Shutdown (Empty) returns (Result) {}
  rpc PokeWatchdog (Empty) returns (Empty) {}
  rpc Ping (Empty) returns (PingReply) {}
}

message Empty {}

message Result {
  bool success = 1;
  string msg = 2;
}

message GetJobsRequest {
}

message GetJobsReply {
  repeated int64 active_bulk_jobs = 1;
}

message GetJobStatusRequest {
  int64 bulk_job_id = 1;
}

message GetJobStatusReply {
  bool finished = 1;
  Result result = 2;

  int32 ops_done = 3;
  int32 total_ops = 4;

  int32 jobs_done = 5;
  int32 jobs_failed = 6;
  int32 total_jobs = 7;

  int32 num_workers = 8;
  int32 failed_workers = 9;
}

message ListTablesResult {
  repeated string tables = 1;
}

message GetTablesParams {
  repeated string tables = 1;
}

message GetVideoMetadataParams {
  repeated string tables = 1;
}

message GetVideoMetadataResult {
  repeated VideoDescriptor videos = 1;
}

message GetTablesResult {
  Result result = 1;
  repeated TableDescriptor tables = 2;
}

message DeleteTablesParams {
  repeated string tables = 1;
}

message NewTableRow {
  repeated bytes columns = 1;
}

message NewTableParams {
  string table_name = 1;
  repeated string columns = 2;
  repeated NewTableRow rows = 3;
}

message WorkerParams {
  string port = 1;
  MachineParameters params = 2;
}

message Registration {
  int32 node_id = 1;
}

message UnregisterWorkerRequest {
  int32 node_id = 1;
}

message WorkerInfo {
  int32 id = 1;
  string address = 2;
}

message RegisteredWorkers {
  repeated WorkerInfo workers = 1;
}

message OpPath {
  string path = 1;
}

message OpRegistration {
  string name = 1;
  bool variadic_inputs = 2;
  repeated Column input_columns = 3;
  repeated Column output_columns = 4;
  bool can_stencil = 5;
  repeated int32 preferred_stencil = 6;
  bool has_bounded_state = 7;
  int32 warmup = 8;
  bool has_unbounded_state = 9;
}

message PythonKernelRegistration {
  string op_name = 1;
  DeviceType device_type = 2;
  bytes kernel_code = 3;
  bytes pickled_config = 4;
  int32 batch_size = 5;
}

message IngestParameters {
  repeated string table_names = 1;
  repeated string video_paths = 2;
  bool inplace = 3;
}

message IngestResult {
  Result result = 1;
  repeated string failed_paths = 2;
  repeated string failed_messages = 3;
}

message ListLoadedOpsReply {
  repeated OpPath registrations = 1;
}

message ListRegisteredOpsReply {
  repeated OpRegistration registrations = 1;
}

message ListRegisteredPythonKernelsReply {
  repeated PythonKernelRegistration registrations = 1;
}

message NextWorkRequest {
  int32 node_id = 1;
  int32 bulk_job_id = 2;
}

message NextWorkReply {
  int64 table_id = 1;
  int64 job_index = 2;
  int64 op_index = 7;
  int64 task_index = 3;
  repeated int64 output_rows = 4 [packed=true];
  bool wait_for_work = 5;
  bool no_more_work = 6;
  bytes data = 8;
}

message FinishedWorkRequest {
  int32 node_id = 1;
  int64 bulk_job_id = 2;

  int64 job_id = 3;
  int64 task_id = 4;
  int64 op_id = 6;
  int64 num_rows = 5;
  bytes data = 7;
}

message FinishedJobRequest {
  int32 node_id = 1;
  int32 bulk_job_id = 2;
  Result result = 3;
}

message BulkJobParameters {
  string job_name = 1;

  repeated Op ops = 2; // Linearized DAG of Ops
  repeated Job jobs = 3;

  // Parameters
  repeated string output_column_names = 19;
  repeated OutputColumnCompression compression = 4;
  MemoryPoolConfig memory_pool_config = 5;
  int32 pipeline_instances_per_node = 6;
  int32 io_packet_size = 7;
  int32 work_packet_size = 8;
  bool profiling = 12;
  int32 load_sparsity_threshold = 13;
  int32 tasks_in_queue_per_pu = 14;
  enum BoundaryCondition {
    REPEAT_EDGE = 0;
    REPEAT_NULL = 1;
    ERROR = 2;
  };
  BoundaryCondition boundary_condition = 15;
  float task_timeout = 16;
  int32 checkpoint_frequency = 17;
  enum ProfilerLevel {
    Debug = 0;
    Info = 1;
    Important = 2;
  };
  ProfilerLevel profiler_level = 18;

  // For master's use only
  DatabaseDescriptor db_meta = 20;
  repeated Column output_columns = 21;
  int32 bulk_job_id = 22;
  int64 base_time = 23;
}

message NewJobReply {
  int32 bulk_job_id = 1;
}

message OpInfoArgs {
  string op_name = 1;
}

message SourceInfoArgs {
  string source_name = 1;
}

message EnumeratorInfoArgs {
  string enumerator_name = 1;
}

message SinkInfoArgs {
  string sink_name = 1;
}

message OpInfo {
  Result result = 1;
  bool variadic_inputs = 2;
  repeated Column input_columns = 3;
  repeated Column output_columns = 4;
  string protobuf_name = 5;
}

message SourceInfo {
  Result result = 1;
  repeated Column output_columns = 4;
  string protobuf_name = 5;
}

message EnumeratorInfo {
  Result result = 1;
  string protobuf_name = 5;
}

message SinkInfo {
  Result result = 1;
  bool variadic_inputs = 2;
  repeated Column input_columns = 3;
  string protobuf_name = 5;
  string stream_protobuf_name = 6;
}

message PingReply {
  int32 node_id = 1;
}
